libtransmission/rpcimpl.cc

```diff
@@ -1696,6 +1706,112 @@ static char const* torrentAdd(tr_session* session, tr_variant* args_in, tr_varia
 ****
 ***/
 
+static char const* groupGet(tr_session* s, tr_variant* args_in, tr_variant* args_out, struct tr_rpc_idle_data* /*idle_data*/)
+{
+    tr_variant* names = nullptr;
+    std::string_view one_name;
+    int names_count = 0;
+
+    if (tr_variantDictFindStrView(args_in, TR_KEY_name, &one_name))
+    {
+        names_count = -1;
+    }
+    else if (tr_variantDictFindList(args_in, TR_KEY_name, &names))
+    {
+        names_count = tr_variantListSize(names);
+    }
+
+    tr_variant* list = tr_variantDictAddList(args_out, TR_KEY_group, 1);
+    for (auto const& [name, group] : s->bandwidth_groups)
+    {
+        bool nameInList = false;
+        if (names_count > 0)
+        {
+            for (int i = 0; !nameInList && i < names_count; i++)
+            {
+                tr_variant* v = tr_variantListChild(names, i);
+                std::string_view l;
+                if (tr_variantIsString(v) && tr_variantGetStrView(v, &l))
+                {
+                    nameInList = name == l;
+                }
+            }
+        }
+
+        if (nameInList || (name == one_name) || (names_count == 0))
+        {
+            bool u = false;
+            bool d = false;
+            uint32_t up = 0;
+            uint32_t down = 0;
+            tr_variant* dict = tr_variantListAddDict(list, 5);
+            tr_bandwidthGroupGetLimits(group, &u, &up, &d, &down);
+            tr_variantDictAddStrView(dict, TR_KEY_name, name);
+            tr_variantDictAddBool(dict, TR_KEY_uploadLimited, u);
+            tr_variantDictAddBool(dict, TR_KEY_downloadLimited, d);
+            tr_variantDictAddInt(dict, TR_KEY_downloadLimit, down);
+            tr_variantDictAddInt(dict, TR_KEY_uploadLimit, up);
+            tr_variantDictAddBool(dict, TR_KEY_honorsSessionLimits, group->areParentLimitsHonored(TR_UP));
+        }
+    }
+
+    return nullptr;
+}
+
+static char const* groupSet(
+    tr_session* session,
+    tr_variant* args_in,
+    tr_variant* /*args_out*/,
+    struct tr_rpc_idle_data* /*idle_data*/)
+{
+    std::string_view name;
+
+    if (!tr_variantDictFindStrView(args_in, TR_KEY_name, &name))
+    {
+        return "No group name given";
+    }
+
+    Bandwidth* group = session->bandwidthGroupFind(name);
+    if (group == nullptr)
+    {
+        return "No such group";
+    }
+
+    bool u = false;
+    bool d = false;
+    uint32_t down = 0;
+    uint32_t up = 0;
+    tr_bandwidthGroupGetLimits(group, &u, &up, &d, &down);
+
+    tr_variantDictFindBool(args_in, TR_KEY_speed_limit_down_enabled, &d);
+    int64_t intVal = 0;
+    if (d && tr_variantDictFindInt(args_in, TR_KEY_speed_limit_down, &intVal))
```

Comment:
we may want to be able to set the speeds even while the limits are off, so maybe the `d &&` here and `u &&` below aren't needed?